src/main/java/net/fabricmc/loader/impl/metadata/ModOriginImpl.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl.metadata;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import net.fabricmc.loader.api.metadata.ModOrigin;
+
+public final class ModOriginImpl implements ModOrigin {
+	private final Kind kind;
+	private List<Path> paths;
+	private String parentModId;
+	private String parentSubLocation;
```

Comment:
Can we have separate interfaces for the two kinds of mod origins so users can perform `instanceof` checks and only call the relevant methods when instance checks pass? This can also make it easier to add new types of mod origins

Reply:
I went with this shape because it allows switch case on the `Kind` enum while still keeping the class count low. Using sub classes would require caller-side downcasts and a bunch of extra classes, which IMO is a bit much for this niche feature and not significantly more elegant.

Reply:
Hmm, then impl-wise, can still split the direct and nested origins to two more classes based on the kind than having to carry useless fields and perform extra checks

Reply:
There are really two choices:
a) have a class for each kind with its respective getters
b) have one class for everything that disallows certain operations a bit like Optional
I went with b due to the niche nature of this API, not sure what else you are suggesting.